sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -231,10 +231,41 @@ object ReorderAssociativeOperator extends Rule[LogicalPlan] {
  * 1. Converts the predicate to false when the list is empty and
  *    the value is not nullable.
  * 2. Removes literal repetitions.
- * 3. Replaces [[In (value, seq[Literal])]] with optimized version
+ * 3. Replaces value IN (x,x+1,x+2..x+n) with x <= value AND value <= x + n
+ * 4. Replaces [[In (value, seq[Literal])]] with optimized version
  *    [[InSet (value, HashSet[Literal])]] which is much faster.
  */
 object OptimizeIn extends Rule[LogicalPlan] {
+  private def isContinousIntegers(nums: Set[Expression]): Boolean = {
+    if (nums.nonEmpty && isInteger(nums.head)) {
+      val (min, max) = getBound(nums)
+      val minL = min.eval(EmptyRow).asInstanceOf[Number].longValue()
+      val maxL = max.eval(EmptyRow).asInstanceOf[Number].longValue()
+       minL + (nums.size - 1) == maxL
+    } else {
+      false
+    }
+  }
+
+  private def isInteger(v: Expression): Boolean =
```

Comment:
Should we handle `ByteType` (tinyint) as well?

Reply:
updated to use IntegralType which covers byte, short, int, long